sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -184,7 +183,8 @@ class Dataset[T] private[sql](
       // to happen right away to let these side effects take place eagerly.
       case p if hasSideEffects(p) =>
         LogicalRDD(queryExecution.analyzed.output, queryExecution.toRdd)(sparkSession)
-      case Union(children) if children.forall(hasSideEffects) =>
+      case Union(children) if children.forall(x =>
+          x.isInstanceOf[InsertIntoTable] || x.isInstanceOf[InsertIntoHadoopFsRelationCommand]) =>
```

Comment:
I see your point. Now we not only support multi-insert, but also union query containing side effect(not intentionally). I don't think it's a big problem and maybe not worth to fix it. cc @hvanhovell @yhuai